token-syncer/src/token_syncer/main.clj

```diff
@@ -0,0 +1,89 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns token-syncer.main
+  (:require [clojure.pprint :as pp]
+            [clojure.string :as str]
+            [clojure.tools.cli :as cli]
+            [clojure.tools.logging :as log]
+            [qbits.jet.client.http :as http]
+            [token-syncer.syncer :as syncer])
+  (:import (org.eclipse.jetty.client HttpClient))
+  (:gen-class))
+
+(defn ^HttpClient http-client-wrapper-factory
+  "Creates an instance of HttpClient with the specified timeout."
+  [{:keys [connection-timeout-ms idle-timeout-ms use-spnego]}]
+  (let [client (http/client {:connect-timeout connection-timeout-ms
+                             :idle-timeout idle-timeout-ms
+                             :follow-redirects? false})
+        _ (.clear (.getContentDecoderFactories client))]
+    {:http-client client
+     :use-spnego use-spnego}))
+
+(defn- setup-exception-handler
+  "Sets up the UncaughtExceptionHandler."
+  []
+  (Thread/setDefaultUncaughtExceptionHandler
+    (reify Thread$UncaughtExceptionHandler
+      (uncaughtException [_ thread throwable]
+        (log/error throwable (str (.getName thread) " threw exception: " (.getMessage throwable)))))))
+
+(defn parse-cli-options
+  "Parses and returns the cli options passed to the token syncer."
+  [args]
+  (let [cli-options [["-c" "--cluster-urls urls" "The semi-colon separated cluster urls"
+                      :default []
+                      :parse-fn #(str/split % #";")
+                      :validate [#(> (count (set %)) 1) "Must provide at least two different semi-colon separated cluster urls"]]
+                     ["-h" "--help"]
+                     ["-i" "--idle-timeout-ms timeout" "The idle timeout in milliseconds"
+                      :default 30000
+                      :parse-fn #(Integer/parseInt %)
+                      :validate [#(< 0 % 300001) "Must be between 0 and 300000"]]
+                     ["-s" "--use-spnego"]
```

Comment:
If I enable this, will it work on a non-spnego cluster?  If not, can it be made to work on a non-spnego cluster (i.e. only try it we actually get challenged)?  Then we can just remove this optionality.

Reply:
Added logic to automatically determine spnego mode in `waiter/make-http-request`